sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/objects.scala

```diff
@@ -132,49 +130,74 @@ object ObjectSerializerPruning extends Rule[LogicalPlan] {
         fields.map(f => collectStructType(f.dataType, structs))
       case ArrayType(elementType, _) =>
         collectStructType(elementType, structs)
+      case MapType(_, valueType, _) =>
+        // Because we can't select a field from struct in key, so we skip key type.
```

Comment:
Technically, , `map_keys` returns key arrays and we can select fields from the keys.
```scala
scala> sql("select keys, keys.a from (select map_keys(map(named_struct('a', 'key1'), 1, named_struct('a', 'key2'), 2)) keys)").show
+----------------+------------+
|            keys|           a|
+----------------+------------+
|[[key1], [key2]]|[key1, key2]|
+----------------+------------+
```

Reply:
This comment is a bit misleading. This should be a TODO as current schema pruning can't work on map keys. Because this isn't the target of this PR, I don't want to address it here. I will fix it in a separate PR.

Reply:
@dongjoon-hyun Created a separate PR at #24179.

Reply:
Thanks a lot, @viirya !